### PR TITLE
test: fix configure check for CXX

### PR DESCRIPTION
**Pull request by hzhou**
## Pull Request Description
* In test/mpi/configure.ac `MPI::Init(NLLL, NULL)` does not work. It should be `MPI::Init()`.
* `HAVE_ROMIO` macro was not working in `mpicxx.h`
* Fix ch3 segfault when sending `MPI_DATATYPE_NULL`
* Fix warnings in `test/mpi/impl/mpich/threadcomm/threadcomm_coll.c`

[skip warnings]
## Future TODO
* [ ] Generate bindings for ROMIO  -- postponed

## Author Checklist
* [x] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [x] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [x] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [x] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

### CODE REVIEW

**Comment by hzhou** [comment]
test:mpich/ch4/ofi
test:mpich/ch3/tcp

**Comment by hzhou** [comment]
test:mpich/ch4/ofi
test:mpich/ch3/tcp

**Comment by hzhou** [comment]
test:mpich/ch4/ofi
test:mpich/ch3/tcp

**Comment by hzhou** [comment]
test:mpich/ch3/tcp
test:mpich/ch4/ofi

**Comment by hzhou** [comment]
test:mpich/ch3/tcp
test:mpich/ch4/ofi

File: src/include/mpi.h.in [hunk 1]
```diff
@@ -998,6 +998,9 @@ typedef struct MPIX_Iov {
 #endif /* MPICH_SUPPRESS_PROTOTYPES */
 
 @HAVE_ROMIO@
+#ifdef HAVE_ROMIO
```
Comment (raffenet) [inline src/include/mpi.h.in:1001]: I'm not sure I love putting this in a header the user will consume. The previous method seems preferable to that, honestly.

Comment (hzhou) [reply src/include/mpi.h.in:1001]: We had an offline discussion. The conclusion is we'll always include mpi-io in mpi.h (and all bindings) and supply a mpi-io stub if `--without-romio` is configured.

Comment (hzhou) [reply src/include/mpi.h.in]: let's postpone the ROMIO binding generation. I have updated the fix to configure time autoconf substitution instead.

**Comment by hzhou** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

**Comment by hzhou** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

**Review by raffenet — APPROVED** [review]

**Pull request merged**